transformer_engine/pytorch/csrc/extensions.cu

```diff
@@ -1767,6 +1767,171 @@ bool userbuf_comm_available() {  // TODO(ksivamani) check on python side
+
+    for (int i = 0; i < Z; ++i) {
+        uint64_t *out = reinterpret_cast<uint64_t*>(my_output + i * load_size * 3);
+        *out = *reinterpret_cast<const uint64_t*>(my_input + i * load_size);
+    }
+}
+
+}  // namespace flash_attention
+
+at::Tensor fa_prepare_fwd(at::Tensor qkvi) {
+    NVTE_CHECK(qkvi.dim() == 4, "Expected 4-dim tensor.");
+    NVTE_CHECK(qkvi.scalar_type() == at::ScalarType::Half ||
+               qkvi.scalar_type() == at::ScalarType::BFloat16);
+    NVTE_CHECK(qkvi.size(3) % flash_attention::load_size == 0);
+    NVTE_CHECK(qkvi.size(3) == flash_attention::load_size);
+
+    // [s, b, n, h * 3] -> [3, b, s, n, h]
+    std::vector<int64_t> shape = {3, qkvi.size(1), qkvi.size(0), qkvi.size(2), qkvi.size(3)};
+    at::Tensor qkv = at::empty(shape, at::CUDA(qkvi.scalar_type()));
+
+    size_t warps = qkvi.size(0) * qkvi.size(1);
+    size_t warps_per_block = flash_attention::block_size / flash_attention::warp_size;
+    size_t blocks = (warps + warps_per_block - 1) / warps_per_block;
+    dim3 grid(blocks, 3);
+    int threads = flash_attention::block_size;
+    if (qkvi.scalar_type() == at::ScalarType::Half) {
+        using dtype = at::Half;
+        flash_attention::prepare_kernel_fwd<dtype><<<grid, threads, 0,
+                                                     at::cuda::getCurrentCUDAStream()>>>(
+            qkvi.data_ptr<dtype>(),
+            qkv.data_ptr<dtype>(),
+            shape[1],
+            shape[2],
+            shape[3],
+            shape[4]);
+    } else {
+        using dtype = at::BFloat16;
+        flash_attention::prepare_kernel_fwd<dtype><<<grid, threads, 0,
+                                                     at::cuda::getCurrentCUDAStream()>>>(
+            qkvi.data_ptr<dtype>(),
+            qkv.data_ptr<dtype>(),
+            shape[1],
+            shape[2],
+            shape[3],
+            shape[4]);
+    }
+
+    return qkv;
+}
+
+at::Tensor fa_prepare_bwd(at::Tensor q, at::Tensor k, at::Tensor v) {
+    NVTE_CHECK(q.is_contiguous());
+    NVTE_CHECK(k.is_contiguous());
+    NVTE_CHECK(v.is_contiguous());
+    NVTE_CHECK(q.dim() == 4, "Expected 4-dim tensor.");
+    NVTE_CHECK(k.dim() == 4, "Expected 4-dim tensor.");
+    NVTE_CHECK(v.dim() == 4, "Expected 4-dim tensor.");
+    NVTE_CHECK(q.scalar_type() == at::ScalarType::Half ||
+               q.scalar_type() == at::ScalarType::BFloat16);
+    NVTE_CHECK(k.scalar_type() == q.scalar_type());
+    NVTE_CHECK(v.scalar_type() == q.scalar_type());
+    NVTE_CHECK(q.size(3) % flash_attention::load_size == 0);
+    NVTE_CHECK(q.size(3) == flash_attention::load_size);
+    NVTE_CHECK(k.size(3) % flash_attention::load_size == 0);
+    NVTE_CHECK(k.size(3) == flash_attention::load_size);
+    NVTE_CHECK(v.size(3) % flash_attention::load_size == 0);
+    NVTE_CHECK(v.size(3) == flash_attention::load_size);
+
+    // 3 x [s, b, n, h] -> [b, s, n, 3 * h]
+
+    std::vector<int64_t> shape = {q.size(1), q.size(0), q.size(2), 3 * q.size(3)};
```

Comment:
```suggestion
    // 3 x [s, b, n, h] -> [b, s, 3 * n, h]

    std::vector<int64_t> shape = {q.size(1), q.size(0), 3 * q.size(2), q.size(3)};
```

Reply:
Same, it is for the last dim split, while the suggestion is for the second to last/